### PR TITLE
Improve `joybus` method structure

### LINKED ISSUE #22: Flush out `Joybus` protocol for allocating and non-allocating use cases
The `joybus` trait only implements `send` and `send_receive`. It seems to reason that it may be useful to provide an allocating and non allocating version of the joybus methods.

Here is some I can imagine for implementations of a useful pipe system to the gamecube controller. These should also be considered for renaming to be more clear on their usage.

1. `send(&send_packet) -> None` can be kept for one way direct messages to the controller
2. `send_receive(&send_packet) -> &receive_packet` can also be kept as the simplest way of sending a message and getting a response
3. `send_receive_no_alloc(&send_packet, &mut receive_packet)` could be added as a way of the usual send receive but without requiring the method to allocate a new packet, and instead filling an existing provided packet

### LINKED ISSUE #23: Use BitArrays instead of standard arrays
currently we use standard arrays instead of bitarrays. this needs to be changed

### PR DESCRIPTION
**Pull request by rhedgeco**
resolves #22 
resolves #23

### CODE REVIEW

**Comment by rhedgeco** [comment]
weird some old master commits got pulled into this. result is still the same though after squash

File: software/fishstick/src/joybus/joybus.rs [hunk 1]
```diff
@@ -1,4 +1,12 @@
...
+pub const JOYBUS_SIZE: usize = 128;
+
+pub type Packet = BitArr!(for JOYBUS_SIZE, in u8, Msb0);
+
 pub trait Joybus {
-    fn send(&self, packet: &[i32]) -> Result<(), ()>;
-    fn send_receive<const S: usize>(&self, packet: &[i32]) -> Result<[i32; S], ()>;
+    fn send(&self, packet: &Packet) -> Result<(), ()>;
+    fn send_receive(&self, packet: &Packet) -> Result<&Packet, ()>;
+    fn send_receive_no_alloc(&self, packet: &Packet, receive_packet: &mut Packet) -> Result<(), ()>;
```
Comment (BryceBeagle) [inline software/fishstick/src/joybus/joybus.rs:11]: For now, we're just assuming input packet length == output packet length, right?

File: software/fishstick/src/joybus/joybus.rs [hunk 2]
```diff
@@ -1,4 +1,12 @@
+use bitvec::prelude::*;
+use bitvec::BitArr;
```
Comment (BryceBeagle) [inline software/fishstick/src/joybus/joybus.rs:2]: Is this resolving #23 as well?

**Review by BryceBeagle — APPROVED** [review]

File: software/fishstick/src/joybus/joybus.rs (hunk 2 above)
Comment (rhedgeco) [reply software/fishstick/src/joybus/joybus.rs:2]: yes this will also resolve #23

File: software/fishstick/src/joybus/joybus.rs (hunk 1 above)
Comment (rhedgeco) [reply software/fishstick/src/joybus/joybus.rs:11]: yes this is what I am thinking for now. we will ultimately need to change the packet length later as well.

also at some point we could switch back to generics and remove the constant altogether

File: software/fishstick/src/joybus/joybus.rs (hunk 2 above)
Comment (BryceBeagle) [reply software/fishstick/src/joybus/joybus.rs:2]: sg. updated the PR description

**Pull request merged**